### PR TITLE
Added `miniforge` support

### PR DESCRIPTION
**Pull request by jiawen**
* `load_conda` learned a new parameter `installer` that can be set to either `miniconda` or `miniforge`, defaulting to `miniconda`.
* Miniforge does not support Python 2.
* Adds MacOS arm64 support via miniforge.
  * miniconda does *not* support Apple Silicon - but this should be ok because the dictionary lookup will fail.

Tested: added tests/miniforge now running in CI.

### CODE REVIEW

**Comment by jiawen** [comment]
Hacked up a quick example for #20 - please take a look. I'm leaving comments in the code itself.

File: example/WORKSPACE
Comment (jiawen) [inline example/WORKSPACE:89]: I will revert this. Just FYI that it doesn't work with miniforge.

File: example/third_party/conda/py3_environment.yml
Comment (jiawen) [inline example/third_party/conda/py3_environment.yml:8]: Probably belongs here? I can pull this into a separate PR.

File: example/WORKSPACE
Comment (jiawen) [inline example/WORKSPACE:32]: Just for debugging. Will revert before merging.

File: conda.bzl
Comment (jiawen) [inline conda.bzl:34]: These are nonsense values copy-pasted from above. They are passed to `rctx.download` but it doesn't seem to be checked?

File: conda.bzl
Comment (jiawen) [inline conda.bzl:32]: @jiawen reminder to self to update this before merging.

File: conda.bzl
Comment (jiawen) [inline conda.bzl:68]: I don't want to overengineer this. Maybe the easiest way is to pass an `attr` to `load_conda_rule = repository_rule(...` with `installer="miniconda"` as the default. You can pass `"miniforge"` as an alternative. We can add `mambaforge` or whatever fancy new thing later.

`_load_conda_impl` or `_download_conda` will check and use the right string as appropriate.

Overengineering would be to use a struct / dataclass / whatnot to hold the collection of strings.

File: example/third_party/conda/py3_environment.yml
Comment (spietras) [reply example/third_party/conda/py3_environment.yml:8]: Nope, it was there there on purpose to show that you can also install packages by `pip`

File: conda.bzl
Comment (spietras) [reply conda.bzl:34]: According to [docs](https://docs.bazel.build/versions/main/skylark/lib/repository_ctx.html#download) the `sha256` passed to `rctx.download` must match the one of the actual file, so I assume the download action should check the integrity itself. If that's not the case then it's kinda weird.

File: example/WORKSPACE
Comment (spietras) [reply example/WORKSPACE:89]: Now I'm wondering if we should drop support for Python 2 altogether 🤔 

It's deprecated anyway and [requiring Python 3 might be a good practice](https://python3statement.org)

File: conda.bzl
Comment (spietras) [reply conda.bzl:68]: Definitely passing `installer` as a parameter with `miniconda` as the default value is a good idea. 

Using structs doesn't really feel like overengineering to me, might even make the code cleaner. But the other way is not bad also. So just follow your intuition 😛

File: example/third_party/conda/py3_environment.yml
Comment (jiawen) [reply example/third_party/conda/py3_environment.yml:8]: SG. Reverted.

File: conda.bzl
Comment (jiawen) [reply conda.bzl:34]: 🤷 . In any case, I updated the hashes.

File: example/WORKSPACE
Comment (jiawen) [reply example/WORKSPACE:89]: I'm happy to drop it :). But maybe not in this PR. I'll leave that up to you. Let's see if tests still pass now that I'm passing `miniconda` as the default.

Do you want me to add a `miniforge` test?

File: example/WORKSPACE
Comment (jiawen) [reply example/WORKSPACE:32]: Done.

File: conda.bzl
Comment (jiawen) [reply conda.bzl]: Done - with just `if / else`.

File: example/WORKSPACE
Comment (jiawen) [inline example/WORKSPACE:89]: Oops, reminder to @jiawen to remove this - pending discussion with @spietras.

File: example/WORKSPACE
Comment (spietras) [reply example/WORKSPACE]: Right, dropping Python 2 should be in another pull request. If the user specifies that he wants to use `miniforge` and also tries to create Python 2 environment then he will get an error, so it's ok. 

As for tests: yes, it's not a must, but having tests for using `miniforge` would be hugely beneficial. See/join the - discussion in #22 about structuring the tests.

Comment (jiawen) [reply example/WORKSPACE]: Thanks - left a comment there.

I'd like to get this in :). Should I migrate `tests` to `tests/miniconda` and add a `tests/miniforge`? I'm happy to do the former in a separate PR.

Comment (spietras) [reply example/WORKSPACE]: Yes, for now, let's have `tests/miniconda` and `tests/miniforge` with duplicated files. Not the best way, but better than no tests at all 😛 I will create another issue with designing and setting up proper test framework

File: conda.bzl
Comment (spietras) [inline conda.bzl:46]: As using `mamba` is orthogonal to using `miniforge` let's just forget about `mambaforge`. If the user wants to use `miniforge` and `mamba` he would specify that he wants to use the `miniforge` installer and that he wants to use `mamba` (after #22). That way we are keeping the decision about installer and about whether to use `conda` or `mamba` separate and independent.

File: conda.bzl
Comment (spietras) [inline conda.bzl:49]: Let's just make single `INSTALLER_FLAGS` constant

File: conda.bzl
Comment (spietras) [inline conda.bzl:158]: Please move it to be the first optional argument

File: utils.bzl
Comment (spietras) [inline utils.bzl:45]: For consistency:
```suggestion
    elif arch in ["arm64"]:
        return "arm64"
```

File: example/WORKSPACE
Comment (jiawen) [reply example/WORKSPACE:89]: Done.

File: example/WORKSPACE
Comment (jiawen) [reply example/WORKSPACE]: Done.

File: conda.bzl
Comment (jiawen) [reply conda.bzl:46]: Reverted comment.

File: conda.bzl
Comment (jiawen) [reply conda.bzl:49]: Done.

File: conda.bzl
Comment (jiawen) [reply conda.bzl:158]: Done.

File: utils.bzl
Comment (jiawen) [reply utils.bzl:45]: Done.

**Comment by jiawen** [comment]
@spietras PTAL. Squash before merging since the history is now super ugly? (I do squash merge on all my projects.)

**Comment by spietras** [comment]
> Squash before merging since the history is now super ugly? (I do squash merge on all my projects.)

Yes, I also prefer having everything squashed on merging to `main`

File: conda.bzl
Comment (spietras) [inline conda.bzl:27]: Move below, just before or after `INSTALLER_DIR`

File: conda.bzl
Comment (spietras) [inline conda.bzl:80]: Formatting:

```suggestion
        fail("Installer must be either miniconda or miniforge.")
```

File: README.md
Comment (spietras) [inline README.md:35]: Please also update `docs/docs/usage/example.md` and `docs/docs/usage/api/md`

**Review by spietras — COMMENTED** [review]
Look's great, couple more little comments

**Review by spietras — COMMENTED** [review]
Look's great, couple more little comments

File: conda.bzl
Comment (spietras) [inline conda.bzl:20]: Also please run [`buildifier`](https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md) at the end if you haven't already

Comment (jiawen) [reply conda.bzl:20]: Done.

**Review by spietras — APPROVED** [review]
Alright, everything looks fine. Great job. 🔥

Thanks for contributing to `rules_conda`.

**Pull request merged**
